src/irmin/key.mli

```diff
@@ -79,9 +79,14 @@
     It's still required that keys be convertible to hashes, as this ensures the
     consistency of Irmin's Merkle tree structure.
 
-    Note that stores that use non-hash-like keys are not content-addressable,
-    since the user can't compute the key of a value given the value itself (and
-    perhaps a value has more than one possible key). *)
+    Stores that use non-hash-like keys are not content-addressable, since the
+    user can't compute the key of a value given the value itself (and perhaps a
+    value has more than one possible key).
+
+    Implementer's note: all key implementations must have a pre-hash
```

Comment:
Is this true in general or just for content-addressable stores?

Reply:
True in general, but less important for content-addressable stores since `key` = `hash` anyway (so presumably `key_t` = `hash_t` too). `Indexable.Of_content_addressable` ensures this is the case.